File: frontend/src/types/analysis.ts
/**
 * Canonical Types for Quantum Pen Flip Predictor
 * Based strictly on docs/TEAM_SHARED_CONTRACT.md and docs/BACKEND.md
 */

export type AnalysisStatus =
  | 'queued'
  | 'processing'
  | 'completed'
  | 'failed'
  | 'tracking_lost';

export interface VideoMetadata {
  fps: number;
  duration: number;
  total_frames: number;
  width?: number;
  height?: number;
}

export interface AnalysisResults {
  rpm: number;
  center_accuracy: number;
  wobble_percent: number;
  average_velocity: number;
  peak_velocity?: number;
  average_acceleration?: number;
  peak_acceleration: number;
  tracking_confidence: number;
  rotations: number;
  overall_score?: number; // Optional technique score if returned by backend
  is_fake_flicker?: boolean;
  verdict?: string;
}

export interface AnalysisFiles {
  trajectory: string;
  overlay_video?: string;
}

export interface AnalysisResponse {
  analysis_id: string;
  status: AnalysisStatus;
  video?: VideoMetadata;
  results?: AnalysisResults;
  files?: AnalysisFiles;
  failure_reason?: string;
  error?: string;
  message?: string;
}

export interface StatusResponse {
  analysis_id: string;
  status: AnalysisStatus;
  progress: number; // 0 to 100
  current_frame?: number;
  total_frames?: number;
  failure_reason?: string;
}

export interface TrajectoryPoint {
  frame: number;
  x: number;
  y: number;
  // Optional enriched tracking data if supplied by backend for HUD overlay
  width?: number;
  height?: number;
  center_x?: number;
  center_y?: number;
  confidence?: number;
  tracking_state?: 'detected' | 'estimated' | 'lost';
}

export interface TrajectoryResponse {
  points: TrajectoryPoint[];
  rotation_center?: {
    x: number;
    y: number;
  };
}

export interface VideoFileValidation {
  isValid: boolean;
  file: File;
  name: string;
  sizeMB: number;
  duration: number;
  width: number;
  height: number;
  fps: number;
  previewUrl: string;
  checks: {
    videoLoaded: boolean;
    highFpsDetected: boolean;
    resolutionSuitable: boolean;
    durationSuitable: boolean;
  };
}

export type TrackingStatePill =
  | 'TRACKING'
  | 'CALIBRATING'
  | 'OCCLUDED'
  | 'LOW CONFIDENCE'
  | 'TRACKING LOST';

File: backend/app/schemas/analysis.py
"""
Pydantic schemas for request and response validation.

Field names are canonical — defined in docs/TEAM_SHARED_CONTRACT.md §8.
Do NOT rename these without updating the shared contract first.
"""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Sub-schemas
# ---------------------------------------------------------------------------


class VideoMetaSchema(BaseModel):
    fps: float
    duration: float
    total_frames: int
    width: Optional[int] = None
    height: Optional[int] = None


class ResultsSchema(BaseModel):
    """Primary and supporting metrics — canonical field names."""

    rpm: float
    center_accuracy: float
    wobble_percent: float
    average_velocity: float
    peak_velocity: float
    average_acceleration: float
    peak_acceleration: float
    tracking_confidence: float
    rotations: int
    overall_score: Optional[float] = None
    is_fake_flicker: Optional[bool] = None
    verdict: Optional[str] = None


class FilesSchema(BaseModel):
    trajectory: str
    overlay_video: str


class TrajectoryPoint(BaseModel):
    frame: int
    x: float
    y: float


class TrajectoryResponse(BaseModel):
    points: List[TrajectoryPoint]


class LowFPSWarning(BaseModel):
    warning: str
    fps: float
    recommended_fps: int = 120


# ---------------------------------------------------------------------------
# Upload / create
# ---------------------------------------------------------------------------


class CreateAnalysisResponse(BaseModel):
    analysis_id: str
    status: str
    warning: Optional[LowFPSWarning] = None


# ---------------------------------------------------------------------------
# Status polling
# ---------------------------------------------------------------------------


class StatusResponse(BaseModel):
    analysis_id: str
    status: str
    progress: int
    current_frame: Optional[int] = None
    total_frames: Optional[int] = None
    failure_reason: Optional[str] = None


# ---------------------------------------------------------------------------
# Full analysis result
# ---------------------------------------------------------------------------


class AnalysisResponse(BaseModel):
    analysis_id: str
    status: str
    video: Optional[VideoMetaSchema] = None
    results: Optional[ResultsSchema] = None
    files: Optional[FilesSchema] = None
    failure_reason: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Errors
# ---------------------------------------------------------------------------


class ErrorResponse(BaseModel):
    error: str
    message: str
